src/Centrifuge.ts

```diff
@@ -1 +1,500 @@
+   * When subscribed to, it emits status updates as it progresses.
+   * When awaited, it returns the final confirmed if successful.
+   * Will additionally prompt the user to switch chains if they're not on the correct chain.
+   *
+   * @example
+   *
+   * ```ts
+   * const tx = this._transact(async function* ({ walletClient, publicClient, chainId, signingAddress, signer }) {
+   *   const permit = yield* doSignMessage('Sign Permit', () => {
+   *     return signPermit(walletClient, signer, chainId, signingAddress, '0xabc...123', '0xdef...456', 1000000n)
+   *   })
+   *   yield* doTransaction('Invest', publicClient, () =>
+   *     walletClient.writeContract({
+   *       address: '0xdef...456',
+   *       abi: ABI.LiquidityPool,
+   *       functionName: 'requestDepositWithPermit',
+   *       args: [1000000n, permit],
+   *     })
+   *   )
+   * }, 1)
+   * tx.subscribe(status => console.log(status))
+   *
+   * // Results in something like the following values being emitted (assuming the user was on the right chain):
+   * // { type: 'SigningMessage', title: 'Sign Permit' }
+   * // { type: 'SignedMessage', title: 'Sign Permit', signed: { ... } }
+   * // { type: 'SigningTransaction', title: 'Invest' }
+   * // { type: 'TransactionPending', title: 'Invest', hash: '0x123...abc' }
+   * // { type: 'TransactionConfirmed', title: 'Invest', hash: '0x123...abc', receipt: { ... } }
+   *
+   * @internal
+   */
+  _transactSequence(
+    transactionCallback: (
+      params: TransactionCallbackParams
+    ) => AsyncGenerator<OperationStatus> | Observable<OperationStatus>,
+    chainId?: number
+  ): Query<OperationStatus> {
+    const targetChainId = chainId ?? this.config.defaultChain
+    const self = this
+    async function* transact() {
+      const { signer } = self
+      if (!signer) throw new Error('Signer not set')
+
+      const publicClient = self.getClient(targetChainId)!
+      const chain = self.getChainConfig(targetChainId)
+      const bareWalletClient = isLocalAccount(signer)
+        ? createWalletClient({ account: signer, chain, transport: http(self.#config.rpcUrls?.[chain.id]) })
+        : createWalletClient({ transport: custom(signer) })
+
+      const [address] = await bareWalletClient.getAddresses()
+      if (!address) throw new Error('No account selected')
+
+      const selectedChain = await bareWalletClient.getChainId()
+      if (selectedChain !== targetChainId) {
+        yield { type: 'SwitchingChain', chainId: targetChainId } as const
+        await bareWalletClient.switchChain({ id: targetChainId })
+      }
+
+      // Re-create the wallet client with the correct chain and account
+      // Saves having to pass `account` and `chain` to every `writeContract` call
+      const walletClient = isLocalAccount(signer)
+        ? (bareWalletClient as WalletClient<any, Chain, AccountType>)
+        : createWalletClient({ account: address, chain, transport: custom(signer) })
```

Comment:
having a hard time understanding why the wallet client has to be recreated, from the code it looks like the wallet client could already initially be set correctly to avoid the `walletClient`/`bareWalletClient` situation. The comment is suggesting that it's being recreated because the chainid may be incorrect but it's using the same chain id as the bareWalletClient

Reply:
it's mainly because of the address. It needs a wallet client to get the signer address, but it needs the address to create the wallet client 😕